_posts/2019-01-09-a-new-recovery-process-for-real-time-pipelines/index.md

```diff
@@ -0,0 +1,302 @@
+
+<!--more-->
+
+<h2 id="overview">1. Overview</h2>
+
+Our current approach to data recovery, [Hadoop Event Recovery][hadoop-recovery], suffers from a few
+issues:
+
+- it's limited to data produced by the batch pipeline
+- you need to code your own recovery almost from scratch in JavaScript
+- you cannot test this JavaScript except by running an actual recovery
+- it doesn't promote reuse: if you run the same recovery twice, you'll need to copy/paste your
+recovery code from one recovery to another
+
+Snowplow Event Recovery aims to tackle most of these issues and make the data recovery process:
+
+- not require any coding for the most common cases
+- extensible when outside the most common cases
+- testable
+- unified across the real-time pipelines (AWS and GCP) and, in the future across all pipelines
+(real-time and batch)
+
+<h2 id="csl">2. Recovery scenarios</h2>
+
+Keeping these goals in mind, we started by thinking about what a recovery is, in essence. For us,
+it is a collection of what we've come to call a recovery scenario.
+
+So, what are recovery scenarios? They are modular and composable processing units that will deal
+with a specific case you want to recover from.
+
+As such, recovery scenarios are, at their essence, made up of two things:
+
+- an error filter, which will serve as a router between bad rows and their appropriate recovery
+scenario(s)
+- a mutation function, which will actually "fix" the payload
+
+For example, if we wanted to recover a set of bad rows consisting of:
+
+- Bad rows that were created due to a missing schema
+- Bad rows that were created due to the payload not conforming to its schema
+- Bad rows that were created due to an enrichment failing
+
+We would use a different recovery scenario for each of them, so three in total:
+
+- a first recovery scenario consisting of:
+  - an error filter checking for missing schema errors
+  - a mutate function which does nothing (assuming the schema has been added since the bad rows
+occurred)
+- a second recovery scenario consisting of:
+  - an error filter checking for payloads not conforming to their schema errors
+  - a mutate function which makes the payloads fit their schema
+- a third recovery scenario consisting of:
+  - an error filter checking for a particular enrichment failing errors
+  - a mutate function which does nothing (assuming the enrichment was misconfigured and we just want
+to rerun it)
+
+<h3 id="out-of-the-box">2.1 Out of the box recovery scenarios</h3>
+
+For the most common recovery scenarios, it makes sense to support them out of the box and not
+require any coding. From the recoveries we've run in the past, we've compiled a list of recovery
+scenarios that are supported out of the box by Snowplow Event Recovery.
+
+In the table below, you can find what this list is made of, it contains:
+
+- the name of the recovery scenario
+- what the mutation function will do
+- an example use case
+- the parameters to this recovery scenario
+
```

Comment:
Hey @BenFradet - plain tables don't render very nicely on the website at all. However, because we use Bootstrap, it is [pretty easy to publish nice tables](https://getbootstrap.com/docs/4.0/content/tables/). Unfortunately it means you can't use markdown to generate the tables. Would you mind redoing the below table using HTML and applying the appropriate Bootstrap formatting please? You can see some examples [here](https://github.com/snowplow/snowplowanalytics.com/blob/master/_posts/2016-03-16-introduction-to-event-data-modeling/index.md).

Reply:
addressed in https://github.com/snowplow/snowplowanalytics.com/pull/1027/commits/a06e8769cb76c0054fd283d8418bc8ad963c1fbd